<!doctype html>
<html class="no-js" lang="en" data-content_root="../../../">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="../../../genindex.html" /><link rel="search" title="Search" href="../../../search.html" /><link rel="next" title="step.scmodel" href="../scmodel/index.html" /><link rel="prev" title="step.utils.plotting" href="../utils/plotting/index.html" />

    <!-- Generated with Sphinx 7.3.7 and Furo 2024.01.29 -->
        <title>step.integration - STEP documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo.css?v=135e06be" />
    <link rel="stylesheet" type="text/css" href="../../../_static/graphviz.css?v=fd3f3429" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/css/custom.css?v=77c81d60" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/furo-extensions.css?v=36a5483c" />
    <link rel="stylesheet" type="text/css" href="../../../_static/css/custom.css?v=77c81d60" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../../index.html"><div class="brand">STEP  documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../../../index.html">
  
  
  <span class="sidebar-brand-text">STEP  documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../../../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../index.html">Installation</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../index.html#usage">Usage</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Usage</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../usages/entries.html">EntryPoints</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../usages/results.html">Results Storage</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../index.html#tutorials">Tutorials</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of Tutorials</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../notebooks/DLFPC.html">Spatial domain identification on 10x Visium DLFPC data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../notebooks/Human_Lymph_Node.html">Performing integrative analysis on scRNA-seq data and 10x Visium data of Human Lymph Node with STEP</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../notebooks/MERFISH.html">Reveal multiple level of biological heterogeneities in Mouse Hypothalamus by MERFISH-seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../notebooks/scRNA-seq.html">Integrate Human Immune Cell datasets with STEP and generate batch-corrected embeddings and gene expressions.</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../notebooks/human_colorectal_cancer.html">Visium HD Human Colorectal Cancer (16 um) cell type clustering &amp; spatial domain identification</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../notebooks/mouse_small_intestine.html">Visium HD Mouse Small Intestine (8 um and 16 um) cell type clustering &amp; spatial domain identification</a></li>
</ul>
</li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="../../index.html#api-reference">API Reference</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle navigation of API Reference</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2 current has-children"><a class="reference internal" href="../index.html">step</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle navigation of step</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l3 has-children"><a class="reference internal" href="../functionality/index.html">step.functionality</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" role="switch" type="checkbox"/><label for="toctree-checkbox-5"><div class="visually-hidden">Toggle navigation of step.functionality</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l4 has-children"><a class="reference internal" href="../functionality/comps/index.html">step.functionality.comps</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" role="switch" type="checkbox"/><label for="toctree-checkbox-6"><div class="visually-hidden">Toggle navigation of step.functionality.comps</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l5"><a class="reference internal" href="../functionality/comps/model_ops/index.html">step.functionality.comps.model_ops</a></li>
<li class="toctree-l5"><a class="reference internal" href="../functionality/comps/trainer/index.html">step.functionality.comps.trainer</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../functionality/base/index.html">step.functionality.base</a></li>
<li class="toctree-l4"><a class="reference internal" href="../functionality/cross_funcmodel/index.html">step.functionality.cross_funcmodel</a></li>
<li class="toctree-l4"><a class="reference internal" href="../functionality/sc_funcmodel/index.html">step.functionality.sc_funcmodel</a></li>
<li class="toctree-l4"><a class="reference internal" href="../functionality/st_funcmodel/index.html">step.functionality.st_funcmodel</a></li>
</ul>
</li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../manager/index.html">step.manager</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" role="switch" type="checkbox"/><label for="toctree-checkbox-7"><div class="visually-hidden">Toggle navigation of step.manager</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l4"><a class="reference internal" href="../manager/save/index.html">step.manager.save</a></li>
</ul>
</li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../models/index.html">step.models</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" role="switch" type="checkbox"/><label for="toctree-checkbox-8"><div class="visually-hidden">Toggle navigation of step.models</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l4"><a class="reference internal" href="../models/clust/index.html">step.models.clust</a></li>
<li class="toctree-l4"><a class="reference internal" href="../models/deconv/index.html">step.models.deconv</a></li>
<li class="toctree-l4"><a class="reference internal" href="../models/distributions/index.html">step.models.distributions</a></li>
<li class="toctree-l4"><a class="reference internal" href="../models/extension/index.html">step.models.extension</a></li>
<li class="toctree-l4"><a class="reference internal" href="../models/knn_map/index.html">step.models.knn_map</a></li>
<li class="toctree-l4"><a class="reference internal" href="../models/nnls/index.html">step.models.nnls</a></li>
<li class="toctree-l4"><a class="reference internal" href="../models/transcriptformer/index.html">step.models.transcriptformer</a></li>
</ul>
</li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../modules/index.html">step.modules</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" role="switch" type="checkbox"/><label for="toctree-checkbox-9"><div class="visually-hidden">Toggle navigation of step.modules</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l4"><a class="reference internal" href="../modules/decoder/index.html">step.modules.decoder</a></li>
<li class="toctree-l4"><a class="reference internal" href="../modules/gnn/index.html">step.modules.gnn</a></li>
<li class="toctree-l4"><a class="reference internal" href="../modules/transformer/index.html">step.modules.transformer</a></li>
</ul>
</li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../utils/index.html">step.utils</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" role="switch" type="checkbox"/><label for="toctree-checkbox-10"><div class="visually-hidden">Toggle navigation of step.utils</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l4 has-children"><a class="reference internal" href="../utils/synthetics/index.html">step.utils.synthetics</a><input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" role="switch" type="checkbox"/><label for="toctree-checkbox-11"><div class="visually-hidden">Toggle navigation of step.utils.synthetics</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l5"><a class="reference internal" href="../utils/synthetics/pseudo_st/index.html">step.utils.synthetics.pseudo_st</a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../utils/dataset/index.html">step.utils.dataset</a></li>
<li class="toctree-l4"><a class="reference internal" href="../utils/gbolt/index.html">step.utils.gbolt</a></li>
<li class="toctree-l4"><a class="reference internal" href="../utils/metrics/index.html">step.utils.metrics</a></li>
<li class="toctree-l4"><a class="reference internal" href="../utils/misc/index.html">step.utils.misc</a></li>
<li class="toctree-l4"><a class="reference internal" href="../utils/plotting/index.html">step.utils.plotting</a></li>
</ul>
</li>
<li class="toctree-l3 current current-page"><a class="current reference internal" href="#">step.integration</a></li>
<li class="toctree-l3"><a class="reference internal" href="../scmodel/index.html">step.scmodel</a></li>
<li class="toctree-l3"><a class="reference internal" href="../stmodel/index.html">step.stmodel</a></li>
</ul>
</li>
</ul>
</li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="edit-this-page">
  <a class="muted-link" href="https://github.com/SGGb0nd/step/edit/main/docs/autoapi/step/integration/index.rst" title="Edit this page">
    <svg aria-hidden="true" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <path d="M4 20h4l10.5 -10.5a1.5 1.5 0 0 0 -4 -4l-10.5 10.5v4" />
      <line x1="13.5" y1="6.5" x2="17.5" y2="10.5" />
    </svg>
    <span class="visually-hidden">Edit this page</span>
  </a>
</div><div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="module-step.integration">
<span id="step-integration"></span><h1>step.integration<a class="headerlink" href="#module-step.integration" title="Link to this heading">¶</a></h1>
<section id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Link to this heading">¶</a></h2>
<div class="table-wrapper colwidths-auto summarytable docutils container" id="id1">
<table class="summarytable docutils align-default" id="id1">
<caption><span class="caption-text">Classes</span><a class="headerlink" href="#id1" title="Link to this table">¶</a></caption>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#step.integration.crossModel" title="step.integration.crossModel"><code class="xref py py-obj docutils literal notranslate"><span class="pre">crossModel</span></code></a></p></td>
<td><p>CrossModalityNrmls is the main class for integrating single-cell RNA-seq and spatial transcriptomics data.</p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="classes">
<h2>Classes<a class="headerlink" href="#classes" title="Link to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="step.integration.crossModel">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">crossModel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sc_adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">anndata.AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">pathlib.Path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">st_adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">anndata.AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">pathlib.Path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">class_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_top_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">geneset_to_use</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">st_sample_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coord_keys</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">('array_row',</span> <span class="pre">'array_col')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">decoder_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'zinb'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_transformed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">module_dim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">decoder_input_dim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hidden_dim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">64</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_modules</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">32</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_checkpoint</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_clip</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">st_batch_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variational</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">logarithm_first</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hvg_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'seurat_v3'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filtered</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dispersion</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'batch-gene'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">device</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#step.integration.crossModel" title="Link to this definition">¶</a></dt>
<dd><p>CrossModalityNrmls is the main class for integrating single-cell RNA-seq and spatial transcriptomics data.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="step.integration.crossModel.adata">
<span class="sig-name descname"><span class="pre">adata</span></span><a class="headerlink" href="#step.integration.crossModel.adata" title="Link to this definition">¶</a></dt>
<dd><p>Single-cell RNA-seq data.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="step.integration.crossModel.st_adata">
<span class="sig-name descname"><span class="pre">st_adata</span></span><a class="headerlink" href="#step.integration.crossModel.st_adata" title="Link to this definition">¶</a></dt>
<dd><p>Spatial transcriptomics data.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="step.integration.crossModel.dataset">
<span class="sig-name descname"><span class="pre">dataset</span></span><a class="headerlink" href="#step.integration.crossModel.dataset" title="Link to this definition">¶</a></dt>
<dd><p>CrossDataset object.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="step.integration.crossModel._functional">
<span class="sig-name descname"><span class="pre">_functional</span></span><a class="headerlink" href="#step.integration.crossModel._functional" title="Link to this definition">¶</a></dt>
<dd><p>CrossModalityNrmls object.</p>
</dd></dl>

<p>Initializes the Integration object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sc_adata</strong> (<em>str</em><em> | </em><em>AnnData</em>) – Single-cell RNA-seq data or path to the file containing the data.</p></li>
<li><p><strong>st_adata</strong> (<em>str</em><em> | </em><em>AnnData</em>) – Spatial transcriptomics data or path to the file containing the data.</p></li>
<li><p><strong>class_key</strong> (<em>str</em>) – Key in the AnnData object that specifies the cell type labels.</p></li>
<li><p><strong>n_top_genes</strong> (<em>Optional</em><em>[</em><em>int</em><em>]</em>) – Number of top genes to select. Default is 2000.</p></li>
<li><p><strong>geneset_to_use</strong> (<em>Optional</em><em>[</em><em>Sequence</em><em>[</em><em>str</em><em>]</em><em>]</em>) – List of genes to use for integration. Default is None.</p></li>
<li><p><strong>st_sample_name</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – Name of the st sample to use. Default is None.</p></li>
<li><p><strong>batch_key</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – Key in the AnnData object that specifies the batch labels. Default is None.</p></li>
<li><p><strong>layer_key</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – Key in the AnnData object that specifies the layer to use. Default is None.</p></li>
<li><p><strong>coord_keys</strong> (<em>Tuple</em><em>[</em><em>str</em><em>, </em><em>str</em><em>]</em>) – Tuple of keys in the AnnData object that specify the spatial coordinates. Default is (‘array_row’, ‘array_col’).</p></li>
<li><p><strong>log_transformed</strong> (<em>bool</em>) – Whether the gene expression data is log-transformed. Default is False.</p></li>
<li><p><strong>module_dim</strong> (<em>int</em>) – Dimension of the latent space. Default is 30.</p></li>
<li><p><strong>decoder_input_dim</strong> (<em>Optional</em><em>[</em><em>int</em><em>]</em>) – Dimension of the decoder input. Default is None.</p></li>
<li><p><strong>hidden_dim</strong> (<em>int</em>) – Dimension of the hidden layers. Default is 64.</p></li>
<li><p><strong>n_modules</strong> (<em>int</em>) – Number of modules in the model. Default is 32.</p></li>
<li><p><strong>model_checkpoint</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – Path to the model checkpoint file. Default is None.</p></li>
<li><p><strong>edge_clip</strong> (<em>int</em>) – Value to clip the edges of the latent space. Default is 2.</p></li>
<li><p><strong>st_batch_key</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – Key in the AnnData object that specifies the batch labels for the st data. Default is None.</p></li>
<li><p><strong>variational</strong> (<em>bool</em>) – Whether to use variational autoencoder. Default is True.</p></li>
<li><p><strong>logarithm_first</strong> (<em>bool</em>) – Whether to apply logarithm to the gene expression data before normalization. Default is False.</p></li>
<li><p><strong>hvg_method</strong> (<em>str</em>) – Method to use for selecting highly variable genes. Default is “seurat_v3”.</p></li>
<li><p><strong>filtered</strong> (<em>bool</em>) – Whether the gene expression data (cells) are filtered. Default is False.</p></li>
<li><p><strong>**kwargs</strong> – Additional keyword arguments to be passed to the CrossModalityNrmls constructor.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Overview</p>
<div class="table-wrapper colwidths-auto summarytable docutils container" id="id2">
<table class="summarytable docutils align-default" id="id2">
<caption><span class="caption-text">Methods</span><a class="headerlink" href="#id2" title="Link to this table">¶</a></caption>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#step.integration.crossModel.cluster" title="step.integration.crossModel.cluster"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cluster</span></code></a>(st_adata, n_clusters, use_rep, key_added, method)</p></td>
<td><p>Cluster the embedding of spatial transcriptomics data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#step.integration.crossModel.sub_cluster" title="step.integration.crossModel.sub_cluster"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sub_cluster</span></code></a>(st_adata, n_clusters, use_rep, pre_key, key_added, method)</p></td>
<td><p>Sub-cluster the clusters of spatial transcriptomics data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#step.integration.crossModel.summarize_domain" title="step.integration.crossModel.summarize_domain"><code class="xref py py-obj docutils literal notranslate"><span class="pre">summarize_domain</span></code></a>(cell_type_names, adata, domain_key, obsm_key, figsize, show, save)</p></td>
<td><p>Summarize the average quantification of cell types in each domain.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#step.integration.crossModel.summarize_single_domain" title="step.integration.crossModel.summarize_single_domain"><code class="xref py py-obj docutils literal notranslate"><span class="pre">summarize_single_domain</span></code></a>(cell_type_names, domain, adata, domain_key, obsm_key, figsize, show, save)</p></td>
<td><p>Summarize the quantification of cell types in each domain.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#step.integration.crossModel.save" title="step.integration.crossModel.save"><code class="xref py py-obj docutils literal notranslate"><span class="pre">save</span></code></a>(path, save_adata)</p></td>
<td><p>Save the model and the data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#step.integration.crossModel.load" title="step.integration.crossModel.load"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load</span></code></a>(path, sc_adata, st_adata, config_name, model_name, st_decoder_name, mixer_name)</p></td>
<td><p><span class="summarylabel">class</span> Load the model and the data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#step.integration.crossModel.spatial_plot" title="step.integration.crossModel.spatial_plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">spatial_plot</span></code></a>(slide, with_images, **kwargs)</p></td>
<td><p>-</p></td>
</tr>
</tbody>
</table>
</div>
<p class="rubric">Members</p>
<dl class="py method">
<dt class="sig sig-object py" id="step.integration.crossModel.cluster">
<span class="sig-name descname"><span class="pre">cluster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">st_adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">anndata.AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_clusters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_rep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'X_smoothed'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'domain'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'kmeans'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#step.integration.crossModel.cluster" title="Link to this definition">¶</a></dt>
<dd><p>Cluster the embedding of spatial transcriptomics data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>st_adata</strong> (<em>Optional</em><em>[</em><em>AnnData</em><em>]</em>) – Spatial transcriptomics data. Default is None, which uses the spatial transcriptomics data provided during initialization.</p></li>
<li><p><strong>n_clusters</strong> (<em>int</em>) – Number of clusters. Default is 3.</p></li>
<li><p><strong>use_rep</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – Key of the embedding to use. Default is None.</p></li>
<li><p><strong>key_added</strong> (<em>str</em>) – Key to add to the obs attribute of the spatial transcriptomics data. Default is “domain”.</p></li>
<li><p><strong>method</strong> (<em>str</em>) – Clustering method, either “kmeans” or “mclust”. Default is “kmeans”.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="step.integration.crossModel.sub_cluster">
<span class="sig-name descname"><span class="pre">sub_cluster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">st_adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">anndata.AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_clusters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_rep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'X_smoothed'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pre_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cluster'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sub_domain'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'kmeans'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#step.integration.crossModel.sub_cluster" title="Link to this definition">¶</a></dt>
<dd><p>Sub-cluster the clusters of spatial transcriptomics data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>st_adata</strong> (<em>Optional</em><em>[</em><em>AnnData</em><em>]</em>) – Spatial transcriptomics data. Default is None, which uses the spatial transcriptomics data provided during initialization.</p></li>
<li><p><strong>n_clusters</strong> (<em>int</em>) – Number of clusters. Default is 3.</p></li>
<li><p><strong>use_rep</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – Key of the embedding to use. Default is None.</p></li>
<li><p><strong>pre_key</strong> (<em>str</em>) – Key in the obs attribute of the spatial transcriptomics data that specifies the clusters. Default is “cluster”.</p></li>
<li><p><strong>key_added</strong> (<em>str</em>) – Key to add to the obs attribute of the spatial transcriptomics data. Default is “sub_domain”.</p></li>
<li><p><strong>method</strong> (<em>str</em>) – Clustering method, either “kmeans” or “mclust”. Default is “kmeans”.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="step.integration.crossModel.summarize_domain">
<span class="sig-name descname"><span class="pre">summarize_domain</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_type_names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">anndata.AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">domain_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'domain'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obsm_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'deconv'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(15,</span> <span class="pre">5)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#step.integration.crossModel.summarize_domain" title="Link to this definition">¶</a></dt>
<dd><p>Summarize the average quantification of cell types in each domain.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cell_type_names</strong> (<em>Sequence</em><em>[</em><em>str</em><em>]</em>) – Names of the cell types.</p></li>
<li><p><strong>adata</strong> (<em>Optional</em><em>[</em><em>AnnData</em><em>]</em>) – Single-cell RNA-seq data. Default is None, which uses the single-cell RNA-seq data provided during initialization.</p></li>
<li><p><strong>domain_key</strong> (<em>str</em>) – Key in the obs attribute of the spatial transcriptomics data that specifies the domain. Default is “domain”.</p></li>
<li><p><strong>obsm_key</strong> (<em>str</em>) – Key in the obsm attribute of the spatial transcriptomics data that specifies the cell type labels. Default is “deconv”.</p></li>
<li><p><strong>figsize</strong> (<em>Tuple</em><em>[</em><em>int</em><em>, </em><em>int</em><em>]</em>) – Figure size. Default is (15, 5).</p></li>
<li><p><strong>show</strong> (<em>bool</em>) – Whether to show the plot. Default is True.</p></li>
<li><p><strong>save</strong> (<em>bool</em>) – Whether to save the plot. Default is True.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="step.integration.crossModel.summarize_single_domain">
<span class="sig-name descname"><span class="pre">summarize_single_domain</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cell_type_names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">domain</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">anndata.AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">domain_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'domain'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obsm_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'deconv'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(15,</span> <span class="pre">5)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#step.integration.crossModel.summarize_single_domain" title="Link to this definition">¶</a></dt>
<dd><p>Summarize the quantification of cell types in each domain.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cell_type_names</strong> (<em>Sequence</em><em>[</em><em>str</em><em>]</em>) – Names of the cell types.</p></li>
<li><p><strong>adata</strong> (<em>Optional</em><em>[</em><em>AnnData</em><em>]</em>) – Single-cell RNA-seq data. Default is None, which uses the single-cell RNA-seq data provided during initialization.</p></li>
<li><p><strong>domain_key</strong> (<em>str</em>) – Key in the obs attribute of the spatial transcriptomics data that specifies the domain. Default is “domain”.</p></li>
<li><p><strong>obsm_key</strong> (<em>str</em>) – Key in the obsm attribute of the spatial transcriptomics data that specifies the cell type labels. Default is “deconv”.</p></li>
<li><p><strong>figsize</strong> (<em>Tuple</em><em>[</em><em>int</em><em>, </em><em>int</em><em>]</em>) – Figure size. Default is (15, 5).</p></li>
<li><p><strong>show</strong> (<em>bool</em>) – Whether to show the plot. Default is True.</p></li>
<li><p><strong>save</strong> (<em>bool</em>) – Whether to save the plot. Default is True.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="step.integration.crossModel.save">
<span class="sig-name descname"><span class="pre">save</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'.'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#step.integration.crossModel.save" title="Link to this definition">¶</a></dt>
<dd><p>Save the model and the data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path</strong> (<em>str</em>) – The path to save the model and the dataset.</p></li>
<li><p><strong>save_adata</strong> (<em>bool</em>) – Whether to save the adata object. Default is False.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="step.integration.crossModel.load">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">pathlib.Path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sc_adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">anndata.AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">pathlib.Path</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">st_adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">anndata.AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">pathlib.Path</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'config.json'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'model.pth'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">st_decoder_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'st_decoder.pth'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mixer_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'mixer.pth'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#step.integration.crossModel.load" title="Link to this definition">¶</a></dt>
<dd><p>Load the model and the data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path</strong> (<em>str</em>) – The path to load the model and the dataset.</p></li>
<li><p><strong>adata</strong> (<em>Optional</em><em>[</em><em>AnnData</em><em>]</em>) – Annotated data object containing the gene expression data.</p></li>
<li><p><strong>filepath</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – Path to a file containing the gene expression data.</p></li>
<li><p><strong>config_name</strong> (<em>str</em>) – The name of the config file.</p></li>
<li><p><strong>model_name</strong> (<em>str</em>) – The name of the backbone model file.</p></li>
<li><p><strong>st_decoder_name</strong> (<em>str</em>) – The name of the st_decoder file.</p></li>
<li><p><strong>mixer_name</strong> (<em>str</em>) – The name of the mixer file.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The scModel object.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="../scmodel/index.html#step.scmodel.scModel" title="step.scmodel.scModel">scModel</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="step.integration.crossModel.spatial_plot">
<span class="sig-name descname"><span class="pre">spatial_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">slide</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">with_images</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#step.integration.crossModel.spatial_plot" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="../scmodel/index.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">step.scmodel</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="../utils/plotting/index.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">step.utils.plotting</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2024, Lounan Li
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">step.integration</a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#classes">Classes</a><ul>
<li><a class="reference internal" href="#step.integration.crossModel"><code class="docutils literal notranslate"><span class="pre">crossModel</span></code></a><ul>
<li><a class="reference internal" href="#step.integration.crossModel.adata"><code class="docutils literal notranslate"><span class="pre">crossModel.adata</span></code></a></li>
<li><a class="reference internal" href="#step.integration.crossModel.st_adata"><code class="docutils literal notranslate"><span class="pre">crossModel.st_adata</span></code></a></li>
<li><a class="reference internal" href="#step.integration.crossModel.dataset"><code class="docutils literal notranslate"><span class="pre">crossModel.dataset</span></code></a></li>
<li><a class="reference internal" href="#step.integration.crossModel._functional"><code class="docutils literal notranslate"><span class="pre">crossModel._functional</span></code></a></li>
<li><a class="reference internal" href="#step.integration.crossModel.cluster"><code class="docutils literal notranslate"><span class="pre">crossModel.cluster()</span></code></a></li>
<li><a class="reference internal" href="#step.integration.crossModel.sub_cluster"><code class="docutils literal notranslate"><span class="pre">crossModel.sub_cluster()</span></code></a></li>
<li><a class="reference internal" href="#step.integration.crossModel.summarize_domain"><code class="docutils literal notranslate"><span class="pre">crossModel.summarize_domain()</span></code></a></li>
<li><a class="reference internal" href="#step.integration.crossModel.summarize_single_domain"><code class="docutils literal notranslate"><span class="pre">crossModel.summarize_single_domain()</span></code></a></li>
<li><a class="reference internal" href="#step.integration.crossModel.save"><code class="docutils literal notranslate"><span class="pre">crossModel.save()</span></code></a></li>
<li><a class="reference internal" href="#step.integration.crossModel.load"><code class="docutils literal notranslate"><span class="pre">crossModel.load()</span></code></a></li>
<li><a class="reference internal" href="#step.integration.crossModel.spatial_plot"><code class="docutils literal notranslate"><span class="pre">crossModel.spatial_plot()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="../../../_static/documentation_options.js?v=5929fcd5"></script>
    <script src="../../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/scripts/furo.js?v=32e29ea5"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=a06671a6"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    </body>
</html>