# STEP: Spatial Transcriptomics Embedding Procedure
[![GitHub Workflow Status](https://img.shields.io/github/workflow/status/SGGb0nd/step-dev/CI)](https://github.com/SGGb0nd/step-dev/actions)

![image](http://docs.3s540lab.cloud/images/STEP_fig_1a.webp)
![image](http://docs.3s540lab.cloud/images/STEP_fig_1b.webp)

## Introduction
STEP, an acronym for Spatial Transcriptomics Embedding Procedure, is a foundation deep learning/AI architecture for the analysis of single-cell RNA (scRNA-seq) and spatially resolved transcriptomics (SRT) data. step introduces a unified approach to stepcess and analyze multiple samples of scRNA-seq data as well as align several sections of SRT data, disregarding location relationships. Furthermore, step conducts integrative analysis across different modalities like scRNA-seq and SRT.

## Key Features

-  Integration of multiple scRNA-seq and single-cell resolution SRT samples to reveal cell-type level heterogeneities
-  Alignment of various SRT data sections contiguous or non-contiguous to identify spatial domains across sections
-  Performance of integrative analysis across modalities (scRNA-seq and SRT) and cell-type deconvolution for the non-single-cell resolution SRT data.

## Installation
```
pip install step-kit
```

require python version 3.10+


## Contribution

We welcome contributions! Please see [`CONTRIBUTING.md`](./CONTRIBUTING.md) for more details!

## License

step is licensed under [LICENSE](./LICENSE)
