<!doctype html>
<html class="no-js" lang="en" data-content_root="../">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="../genindex.html" /><link rel="search" title="Search" href="../search.html" /><link rel="next" title="Reveal multiple level of biological heterogeneities in Mouse Hypothalamus by MERFISH-seq" href="MERFISH.html" /><link rel="prev" title="Spatial domain identification on 10x Visium DLFPC data" href="DLFPC.html" />

    <!-- Generated with Sphinx 7.2.6 and Furo 2024.01.29 -->
        <title>Performing integrative analysis on scRNA-seq data and 10x Visium data of Human Lymph Node with STEP - STEP documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?v=135e06be" />
    <link rel="stylesheet" type="text/css" href="../_static/graphviz.css?v=eafc0fe6" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/nbsphinx-code-cells.css" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?v=36a5483c" />
    <link rel="stylesheet" type="text/css" href="../_static/css/custom.css?v=f8d6e08a" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">STEP  documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  
  <span class="sidebar-brand-text">STEP  documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../autoapi/index.html">Installation</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="../autoapi/index.html#tutorials">Tutorials</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Tutorials</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="DLFPC.html">Spatial domain identification on 10x Visium DLFPC data</a></li>
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">Performing integrative analysis on scRNA-seq data and 10x Visium data of Human Lymph Node with STEP</a></li>
<li class="toctree-l2"><a class="reference internal" href="MERFISH.html">Reveal multiple level of biological heterogeneities in Mouse Hypothalamus by MERFISH-seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="scRNA-seq.html">Integrate Human Immune Cell datasets with STEP and generate batch-corrected embeddings and gene expressions.</a></li>
<li class="toctree-l2"><a class="reference internal" href="human_colorectal_cancer.html">Visium HD Human Colorectal Cancer (16 um) cell type clustering &amp; spatial domain identification</a></li>
<li class="toctree-l2"><a class="reference internal" href="mouse_small_intestine.html">Visium HD Mouse Small Intestine (8 um and 16 um) cell type clustering &amp; spatial domain identification</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../autoapi/index.html#api-reference">API Reference</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of API Reference</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../autoapi/step/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle navigation of step</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3 has-children"><a class="reference internal" href="../autoapi/step/functionality/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.functionality</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle navigation of step.functionality</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l4 has-children"><a class="reference internal" href="../autoapi/step/functionality/comps/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.functionality.comps</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" role="switch" type="checkbox"/><label for="toctree-checkbox-5"><div class="visually-hidden">Toggle navigation of step.functionality.comps</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l5"><a class="reference internal" href="../autoapi/step/functionality/comps/model_ops/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.functionality.comps.model_ops</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../autoapi/step/functionality/comps/trainer/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.functionality.comps.trainer</span></code></a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/functionality/base/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.functionality.base</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/functionality/cross_funcmodel/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.functionality.cross_funcmodel</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/functionality/sc_funcmodel/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.functionality.sc_funcmodel</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/functionality/st_funcmodel/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.functionality.st_funcmodel</span></code></a></li>
</ul>
</li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../autoapi/step/manager/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.manager</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" role="switch" type="checkbox"/><label for="toctree-checkbox-6"><div class="visually-hidden">Toggle navigation of step.manager</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/manager/save/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.manager.save</span></code></a></li>
</ul>
</li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../autoapi/step/models/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.models</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" role="switch" type="checkbox"/><label for="toctree-checkbox-7"><div class="visually-hidden">Toggle navigation of step.models</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/models/clust/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.models.clust</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/models/deconv/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.models.deconv</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/models/distributions/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.models.distributions</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/models/extension/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.models.extension</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/models/geneformer/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.models.geneformer</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/models/knn_map/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.models.knn_map</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/models/nnls/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.models.nnls</span></code></a></li>
</ul>
</li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../autoapi/step/modules/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.modules</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" role="switch" type="checkbox"/><label for="toctree-checkbox-8"><div class="visually-hidden">Toggle navigation of step.modules</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/modules/decoder/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.modules.decoder</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/modules/gnn/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.modules.gnn</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/modules/transformer/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.modules.transformer</span></code></a></li>
</ul>
</li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../autoapi/step/utils/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.utils</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" role="switch" type="checkbox"/><label for="toctree-checkbox-9"><div class="visually-hidden">Toggle navigation of step.utils</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l4 has-children"><a class="reference internal" href="../autoapi/step/utils/synthetics/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.utils.synthetics</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" role="switch" type="checkbox"/><label for="toctree-checkbox-10"><div class="visually-hidden">Toggle navigation of step.utils.synthetics</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l5"><a class="reference internal" href="../autoapi/step/utils/synthetics/pseudo_st/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.utils.synthetics.pseudo_st</span></code></a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/utils/dataset/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.utils.dataset</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/utils/gbolt/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.utils.gbolt</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/utils/metrics/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.utils.metrics</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/utils/misc/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.utils.misc</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/utils/plotting/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.utils.plotting</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../autoapi/step/integration/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.integration</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../autoapi/step/scmodel/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.scmodel</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../autoapi/step/stmodel/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.stmodel</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="edit-this-page">
  <a class="muted-link" href="https://github.com/SGGb0nd/step/edit/main/docs/notebooks/Human_Lymph_Node.ipynb" title="Edit this page">
    <svg aria-hidden="true" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <path d="M4 20h4l10.5 -10.5a1.5 1.5 0 0 0 -4 -4l-10.5 10.5v4" />
      <line x1="13.5" y1="6.5" x2="17.5" y2="10.5" />
    </svg>
    <span class="visually-hidden">Edit this page</span>
  </a>
</div><div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <style>
    .nbinput .prompt,
    .nboutput .prompt {
        display: none;
    }
</style><section id="Performing-integrative-analysis-on-scRNA-seq-data-and-10x-Visium-data-of-Human-Lymph-Node-with-STEP">
<h1>Performing integrative analysis on scRNA-seq data and 10x Visium data of Human Lymph Node with STEP<a class="headerlink" href="#Performing-integrative-analysis-on-scRNA-seq-data-and-10x-Visium-data-of-Human-Lymph-Node-with-STEP" title="Link to this heading">#</a></h1>
<p>This tutorial showcases the workflow of integrating two modalities with STEP where they are firstly co-embeded into a shared latent space; then the spatial domains(and subdomains) are identified based on the obtained embedding; subsequently, the cell-type composition of each spot is inferred based on the identified domains and according to the embedding in the shared latent space.</p>
<section id="Loading-packages">
<h2>Loading packages<a class="headerlink" href="#Loading-packages" title="Link to this heading">#</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import torch
import scanpy as sc
import numpy as np
import seaborn as sns
import pandas as pd
import matplotlib.pyplot as plt
import os

from step import crossModel
from step.utils.data import plot_posterior_mu_vs_data

sc_file = &#39;./results/lymph/human_lymph_integrated.h5ad&#39;
sc_pth = &#39;./results/lymph/human_lymph_integrated.pth&#39;
st_pth = &#39;./results/lymph/human_lymph_st_decoder.pth&#39;
st_file = &#39;./results/lymph/human_lymph_filtered.h5ad&#39;

if not os.path.exists(&#39;./results/lymph&#39;):
    os.makedirs(&#39;./results/lymph&#39;)
</pre></div>
</div>
</div>
</section>
<section id="Load-data">
<h2>Load data<a class="headerlink" href="#Load-data" title="Link to this heading">#</a></h2>
<p>This data loading part refers to the <a class="reference external" href="https://cell2location.readthedocs.io/en/latest/notebooks/cell2location_tutorial.html#Mapping-human-lymph-node-cell-types-to-10X-Visium-with-Cell2location">tutorial of cell2location</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_ref = sc.read(
    f&#39;./data/lymph_sc.h5ad&#39;,
    backup_url=&#39;https://cell2location.cog.sanger.ac.uk/paper/integrated_lymphoid_organ_scrna/RegressionNBV4Torch_57covariates_73260cells_10237genes/sc.h5ad&#39;
)
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_vis = sc.datasets.visium_sge(sample_id=&quot;V1_Human_Lymph_Node&quot;)
adata_vis.var_names_make_unique()
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_vis.var[&#39;SYMBOL&#39;] = adata_vis.var_names
adata_vis.var.set_index(&#39;gene_ids&#39;, drop=True, inplace=True)
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_ref.var[&#39;SYMBOL&#39;] = adata_ref.var.index
# rename &#39;GeneID-2&#39; as necessary for your data
adata_ref.var.set_index(&#39;GeneID-2&#39;, drop=True, inplace=True)

# delete unnecessary raw slot (to be removed in a future version of the tutorial)
del adata_ref.raw
</pre></div>
</div>
</div>
</section>
<section id="Initialize-the-STEP-instance">
<h2>Initialize the STEP instance<a class="headerlink" href="#Initialize-the-STEP-instance" title="Link to this heading">#</a></h2>
<p>Initialization of STEP instance requires set the following parameters: - <code class="docutils literal notranslate"><span class="pre">sc_adata</span></code>: Anndata of scRNA-seq data - <code class="docutils literal notranslate"><span class="pre">st_adata</span></code>: Anndata of spatial data - <code class="docutils literal notranslate"><span class="pre">n_top_genes</span></code> or <code class="docutils literal notranslate"><span class="pre">geneset</span></code>: using hvgs or specified geneset - <code class="docutils literal notranslate"><span class="pre">batch_key</span></code>: a column name of <code class="docutils literal notranslate"><span class="pre">sc_adata.obs</span></code> specifying the batches in scRNA-seq data - <code class="docutils literal notranslate"><span class="pre">class_key</span></code>: a column name of <code class="docutils literal notranslate"><span class="pre">sc_adata.obs</span></code> specifying the cell-type annoations in scRNA-seq data - <code class="docutils literal notranslate"><span class="pre">decoder_type</span></code>: distribution(zinb or nb, default zinb) to model gene expression
data.</p>
<p>Other parameters are about model settings: - <code class="docutils literal notranslate"><span class="pre">module_dim</span></code>: dimension of final embedding - <code class="docutils literal notranslate"><span class="pre">hidden_dim</span></code>: dimension of hidden layers - <code class="docutils literal notranslate"><span class="pre">n_moudles</span></code>: number of gene modules - <code class="docutils literal notranslate"><span class="pre">dec_norm</span></code>: normalization layer used in decoder, - ‘layer’: LayerNorm - ‘batch’: BatchNorm - None: not using any normalization layer - <code class="docutils literal notranslate"><span class="pre">n_dec_hid_layers</span></code>: number of hidden layers used in decoder.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>proc = crossModel(
    sc_adata=adata_ref,
    st_adata=adata_vis,
    n_top_genes=2000,
    layer_key=None,
    batch_key=&#39;Sample&#39;,
    class_key=&#39;Subset&#39;,
    module_dim=30,
    hidden_dim=64,
    n_modules=32,
    decoder_type=&#39;nb&#39;,
    dec_norm=None,
    n_dec_hid_layers=1,
    # model_checkpoint=sc_pth,
)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
not log_transformed
Adding count data to layer &#39;counts&#39;
Dataset Done
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>proc.adata.obs[&#39;Sample&#39;].value_counts()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
BCP005_Total              13323
BCP006_Total               4531
4861STDY7462253            4286
4861STDY7462255            4283
V1_Human_Lymph_Node        4035
4861STDY7528597            3888
BCP009_Total               3867
4861STDY7462256            3852
4861STDY7528600            3594
4861STDY7528599            3172
4861STDY7135914            3126
4861STDY7462254            3085
4861STDY7528598            2961
4861STDY7135913            2941
BCP008_Total               2914
Pan_T7935494               2876
BCP003_Total               2083
BCP002_Total               2002
Human_colon_16S8000484     1308
4861STDY7208413            1228
4861STDY7208412            1130
Human_colon_16S7255678     1125
BCP004_Total                985
Human_colon_16S7255677      700
Name: Sample, dtype: int64
</pre></div></div>
</div>
</section>
<section id="Co-embedding-two-modalities">
<h2>Co-embedding two modalities<a class="headerlink" href="#Co-embedding-two-modalities" title="Link to this heading">#</a></h2>
<p>In this process, the goal is to extract the information from transcripts, also eliminate the technical/batch effects in the meantime, so the spatial data is not differentiated from scRNA-seq data by ignoring the spatial locations.</p>
<p>The following running parameters</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>tune_epochs=200,
need_anchors=False,
beta1=1e-2
</pre></div>
</div>
<p>are somewhat special: only when STEP is required to refine the model with cell-type annotations, where <code class="docutils literal notranslate"><span class="pre">need_anchors=True</span></code> is set, the <code class="docutils literal notranslate"><span class="pre">tune_epochs</span></code> is valid and control the number of iterations of the refine process. The <code class="docutils literal notranslate"><span class="pre">beta1</span></code> is the coefficient of KL-divergence, the adjustment of this parameter can result in the change of the degree of embedding’s “dispersion”: higher values result in the lower degree of “dispersion”.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>proc.integrate(epochs=400,
               batch_size=2048,
               split_rate=0.2,
               tune_epochs=200,
               need_anchors=False,
               beta1=1e-2)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Performing category random split
Training size for 4861STDY7135913: 2352
Training size for 4861STDY7135914: 2500
Training size for 4861STDY7208412: 904
Training size for 4861STDY7208413: 982
Training size for 4861STDY7462253: 3428
Training size for 4861STDY7462254: 2468
Training size for 4861STDY7462255: 3426
Training size for 4861STDY7462256: 3081
Training size for 4861STDY7528597: 3110
Training size for 4861STDY7528598: 2368
Training size for 4861STDY7528599: 2537
Training size for 4861STDY7528600: 2875
Training size for BCP002_Total: 1601
Training size for BCP003_Total: 1666
Training size for BCP004_Total: 788
Training size for BCP005_Total: 10658
Training size for BCP006_Total: 3624
Training size for BCP008_Total: 2331
Training size for BCP009_Total: 3093
Training size for Human_colon_16S7255677: 560
Training size for Human_colon_16S7255678: 900
Training size for Human_colon_16S8000484: 1046
Training size for Pan_T7935494: 2300
Training size for V1_Human_Lymph_Node: 3228
train size: 61826
valid size: 15469
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
 86%|████████▌ | 344/400 [27:23&lt;04:27,  4.78s/epoch, kl_loss=1.198, recon_loss=1147.656, val_kl_loss=1.229/0.233, val_recon_loss=1147.362/1143.838]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Early Stopping triggered
EarlyStopping counter: 30 out of 30
EarlyStopping counter: 344 out of 10
</pre></div></div>
</div>
<p>Examine QC plots.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>proc.impute(layer_key=&#39;counts&#39;, qc=True)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Human_Lymph_Node_19_0.png" class="no-scaled-link" src="../_images/notebooks_Human_Lymph_Node_19_0.png" style="width: 269px; height: 303px;" />
</div>
</div>
<section id="Get-and-check-the-co-embedding-results">
<h3>Get and check the co-embedding results<a class="headerlink" href="#Get-and-check-the-co-embedding-results" title="Link to this heading">#</a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = proc.adata
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pp.neighbors(adata, use_rep=&#39;X_rep&#39;)
sc.tl.umap(adata)
</pre></div>
</div>
</div>
<p>As we can see, two modalities manage to “mix” with each other in the left figure. Moreover, in the right figure, we can see that the structure of cell-types is totally preserved, indicating the integration(co-embedding) is biologically meaningful and the technical/batch effects are successfully eliminated.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.umap(adata, color=[&#39;modality&#39;, &#39;Subset&#39;], ncols=2)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Human_Lymph_Node_24_0.png" class="no-scaled-link" src="../_images/notebooks_Human_Lymph_Node_24_0.png" style="width: 636px; height: 301px;" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>torch.save(proc._functional.model.state_dict(), sc_pth)
</pre></div>
</div>
</div>
</section>
</section>
<section id="Zonation-after-integration">
<h2>Zonation after integration<a class="headerlink" href="#Zonation-after-integration" title="Link to this heading">#</a></h2>
<p>Zonation, i.e. identification of spatial domains, is then conducted on the obtained embedding in the previous co-embedding process.</p>
<div class="admonition note">
<p><strong>Note:</strong> This process required embedding produced in the previous process, so if the adata is not saved while the model(paramters) is saved, load the step instance <code class="docutils literal notranslate"><span class="pre">proc</span></code> and use <code class="docutils literal notranslate"><span class="pre">proc.add_embed(key_added='&lt;your</span> <span class="pre">desired</span> <span class="pre">key</span> <span class="pre">name&gt;')</span></code></p>
</div>
<p>Training a spatial smoother(GCN) to spatially smooth the embedding extracted solely from transcprits. This training stage only involves the spatial data and is achieved in two sub-stage: 1. (Optional) Traning a spatial-data-specific decoder for the concentration on spatial data. This sub-stage still doesn’t utilize the spatial context, and is controlled by the following parameters: - <code class="docutils literal notranslate"><span class="pre">epochs</span></code>: #epochs, - <code class="docutils literal notranslate"><span class="pre">batch_size</span></code>: batch size, - <code class="docutils literal notranslate"><span class="pre">n_dec_layers</span></code>: #layers in decoder, - <code class="docutils literal notranslate"><span class="pre">split_rate</span></code>: the
proportion of data used as validation - <code class="docutils literal notranslate"><span class="pre">use_st_decoder</span></code>: whether to run this sub-stage, - <code class="docutils literal notranslate"><span class="pre">decoder_type</span></code>: the distribution to fit count data. 2. Training the spatial smoother to smooth the embedding of spatial data for zonation. This sub-stage is controlled by the following parameters: - <code class="docutils literal notranslate"><span class="pre">n_glayers</span></code>: #graph convolutional layers - <code class="docutils literal notranslate"><span class="pre">smooth_epochs</span></code>: #epochs in this sub-stage - <code class="docutils literal notranslate"><span class="pre">tune_lr</span></code>: learning rate to tune the st-decoder trained in the previous sub-stage, <code class="docutils literal notranslate"><span class="pre">None</span></code> means not tuning it.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>proc.generate_domains(epochs=800,
                      batch_size=1024,
                      n_dec_layers=1,
                      n_glayers=1,
                      split_rate=0.,
                      tune_lr=None,
                      smooth_epochs=800,
                      use_st_decoder=True,
                      decoder_type=&#39;nb&#39;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
==============Sub-Dataset Info==============
Batch key: Sample
Class key: Subset
Number of Batches: 24
Number of Classes: 34
Gene Expr: (4035, 2000)
Rep: (4035, 64)
Batch Label: (4035,)
============================================
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 800/800 [00:58&lt;00:00, 13.63epoch/s, recon_loss=4178.889, val_recon_loss=-]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
StDecoder trained
Start training smoother
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 800/800 [00:47&lt;00:00, 16.79epoch/s, contrast_loss=5.846, recon_loss=4188.098, val_contrast_loss=-, val_recon_loss=-]
</pre></div></div>
</div>
<p>In defualt, the smoothed embedding will be stored in <code class="docutils literal notranslate"><span class="pre">proc.st_adata.obsm['X_smoothed']</span></code>. Here, we use KMeans to cluster the smoothed embedding and obtain 8 spatial domains.</p>
<p>Subdomains is identified by a further clustering based on spatial domains. We identifies 3 subdomains in each domain.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>proc.cluster(use_rep=&#39;X_smoothed&#39;, n_clusters=8)
sc.pl.spatial(proc.st_adata, color=[&#39;domain&#39;], size=1.3,)

proc.sub_cluster(proc.st_adata,
                 pre_key=&#39;domain&#39;,
                 use_rep=&#39;X_smoothed&#39;,
                 n_clusters=3)
sc.pl.spatial(proc.st_adata, color=[&#39;sub_domain&#39;], size=1.3,)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Human_Lymph_Node_33_0.png" class="no-scaled-link" src="../_images/notebooks_Human_Lymph_Node_33_0.png" style="width: 292px; height: 297px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Human_Lymph_Node_33_1.png" class="no-scaled-link" src="../_images/notebooks_Human_Lymph_Node_33_1.png" style="width: 370px; height: 297px;" />
</div>
</div>
<p>Examine the QC plots with original embedding and smoothed embedding through a same decoder(st-decoder if <code class="docutils literal notranslate"><span class="pre">use_st_decoder</span></code> is set previously)</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>proc.st_impute(qc=True, smooth=False)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Human_Lymph_Node_35_0.png" class="no-scaled-link" src="../_images/notebooks_Human_Lymph_Node_35_0.png" style="width: 276px; height: 303px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>proc.st_impute(qc=True, smooth=True)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Human_Lymph_Node_36_0.png" class="no-scaled-link" src="../_images/notebooks_Human_Lymph_Node_36_0.png" style="width: 287px; height: 304px;" />
</div>
</div>
</section>
<section id="Cell-type-deconvolution-with-obtained-embedding">
<h2>Cell-type deconvolution with obtained embedding<a class="headerlink" href="#Cell-type-deconvolution-with-obtained-embedding" title="Link to this heading">#</a></h2>
<p>STEP performs deconvolution at domain-wise since the spots in the same spatial domain share a similar cell-type composition. So this process requires two features: 1. Embedding of scRNA-seq and spatial data in the shared latent space obtained in the co-embedding process. 2. Spatial domains or subdomains stored in <code class="docutils literal notranslate"><span class="pre">proc.st_adata.obs[&lt;domain_key&gt;]</span></code>.</p>
<p>The embedding is mainly used to infer the existance of a cell-type in a spot by evaluating the similarity in latent space between the corresponding embedding. Specifically, the embedding of scRNA-seq data is averaged by cell-types, denoted as <code class="docutils literal notranslate"><span class="pre">anchors</span></code>, and the similarity is measured by a <em>sparse attention mechanism</em> where “sparse” is to ensure the sparsity of cell-type occurence in a spot. To consider the individual effect in each spot, <code class="docutils literal notranslate"><span class="pre">anchors</span></code> are refined individually for each spot
through several(default 2) transformer encoders(can be disabled by setting <code class="docutils literal notranslate"><span class="pre">hard_anchors=True</span></code>). And to bound the refined <code class="docutils literal notranslate"><span class="pre">anchors</span></code>, we add a contrastive loss between refined <code class="docutils literal notranslate"><span class="pre">anchors</span></code> and original <code class="docutils literal notranslate"><span class="pre">anchors</span></code> as a regularization where the parameter<code class="docutils literal notranslate"><span class="pre">T</span></code> is the temperature in the contrastive loss.</p>
<p>The actual cell-type composition in a spot are inferred at transcript-wise by scaling the simlarity matrix inferred in the latent space. This is done by minimizing the negative log-likelihood with the “convolution” of estimated cell-type composition and cell-type signatures. Here the signatures can be obtained by averging the raw counts(set <code class="docutils literal notranslate"><span class="pre">use_raw=True</span></code>) or the decoder-imputed counts(set <code class="docutils literal notranslate"><span class="pre">use_raw=False</span></code>).</p>
<p>For the balance between multiple losses, use parameter <code class="docutils literal notranslate"><span class="pre">rate</span></code> to scale the two majority reconstruction lossess.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>proc.st_adata.obs[&#39;sub_domain_codes&#39;] = proc.st_adata.obs[&#39;sub_domain&#39;].cat.codes

proc.deconv(epochs=2000,
            domain_key=&#39;sub_domain_codes&#39;,
            domain_wise=True,
            n_glayers=None,
            use_raw=False,
            hard_anchors=False,
            T=0.1,
            rate=0.01,)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 2000/2000 [05:14&lt;00:00,  6.36epoch/s, div=2.186, loss_ori=32.250, recon_loss=2080.224, sc_consistency=0.709, val_div=-, val_loss_ori=-, val_recon_loss=-, val_sc_consistency=-]
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>st_adata = proc.st_adata

celltypes = proc.adata.obs[&#39;Subset&#39;].cat.categories
prop_list = celltypes.to_list()
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>st_adata
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 4035 × 2000
    obs: &#39;in_tissue&#39;, &#39;array_row&#39;, &#39;array_col&#39;, &#39;batch_st&#39;, &#39;Sample&#39;, &#39;modality&#39;, &#39;n_genes&#39;, &#39;domain&#39;, &#39;sub_domain&#39;, &#39;sub_domain_codes&#39;, &#39;B_Cycling&#39;, &#39;B_GC_DZ&#39;, &#39;B_GC_LZ&#39;, &#39;B_GC_prePB&#39;, &#39;B_IFN&#39;, &#39;B_activated&#39;, &#39;B_mem&#39;, &#39;B_naive&#39;, &#39;B_plasma&#39;, &#39;B_preGC&#39;, &#39;DC_CCR7+&#39;, &#39;DC_cDC1&#39;, &#39;DC_cDC2&#39;, &#39;DC_pDC&#39;, &#39;Endo&#39;, &#39;FDC&#39;, &#39;ILC&#39;, &#39;Macrophages_M1&#39;, &#39;Macrophages_M2&#39;, &#39;Mast&#39;, &#39;Monocytes&#39;, &#39;NK&#39;, &#39;NKT&#39;, &#39;T_CD4+&#39;, &#39;T_CD4+_TfH&#39;, &#39;T_CD4+_TfH_GC&#39;, &#39;T_CD4+_naive&#39;, &#39;T_CD8+_CD161+&#39;, &#39;T_CD8+_cytotoxic&#39;, &#39;T_CD8+_naive&#39;, &#39;T_TIM3+&#39;, &#39;T_TfR&#39;, &#39;T_Treg&#39;, &#39;VSMC&#39;
    var: &#39;feature_types&#39;, &#39;genome&#39;, &#39;SYMBOL&#39;, &#39;mt_gene&#39;
    uns: &#39;spatial&#39;, &#39;domain_colors&#39;, &#39;sub_domain_colors&#39;
    obsm: &#39;spatial&#39;, &#39;mt&#39;, &#39;X_rep_tsfmr&#39;, &#39;X_smoothed&#39;, &#39;deconv&#39;, &#39;deconv_abundance&#39;
    layers: &#39;counts&#39;, &#39;st_expected_counts&#39;
</pre></div></div>
</div>
<section id="Visualize-the-inferred-cell-type-composition">
<h3>Visualize the inferred cell-type composition<a class="headerlink" href="#Visualize-the-inferred-cell-type-composition" title="Link to this heading">#</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>st_adata.obs[prop_list] = st_adata.obsm[&#39;deconv&#39;]
sc.pl.spatial(st_adata, size=1.3, color=prop_list,)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Human_Lymph_Node_43_0.png" class="no-scaled-link" src="../_images/notebooks_Human_Lymph_Node_43_0.png" style="width: 1263px; height: 2528px;" />
</div>
</div>
</section>
<section id="Perform-spatially-DE-genes-anlaysis">
<h3>Perform spatially DE genes anlaysis<a class="headerlink" href="#Perform-spatially-DE-genes-anlaysis" title="Link to this heading">#</a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_vis.obs[&#39;domain&#39;] = st_adata.obs[&#39;domain&#39;].astype(str).astype(&#39;category&#39;)
adata_vis.obsm[&#39;X_smoothed&#39;] = st_adata.obsm[&#39;X_smoothed&#39;]

sc.tl.dendrogram(adata_vis, use_rep=&#39;X_smoothed&#39;, groupby=&#39;domain&#39;)
sc.tl.rank_genes_groups(adata_vis, groupby=&#39;domain&#39;, method=&#39;wilcoxon&#39;, use_raw=False)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.rank_genes_groups_matrixplot(adata_vis, groupby=&#39;domain&#39;, groups=[&#39;1&#39;, &#39;4&#39;, &#39;8&#39;], values_to_plot=&#39;logfoldchanges&#39;, cmap=&#39;RdBu_r&#39;, vmin=-2, vmax=2)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: Groups are not reordered because the `groupby` categories and the `var_group_labels` are different.
categories: 1, 2, 3, etc.
var_group_labels: 1, 4, 8
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;mainplot_ax&#39;: &lt;AxesSubplot:&gt;,
 &#39;group_extra_ax&#39;: &lt;AxesSubplot:&gt;,
 &#39;gene_group_ax&#39;: &lt;AxesSubplot:&gt;,
 &#39;color_legend_ax&#39;: &lt;AxesSubplot:title={&#39;center&#39;:&#39;log fold change&#39;}&gt;}
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Human_Lymph_Node_46_2.png" class="no-scaled-link" src="../_images/notebooks_Human_Lymph_Node_46_2.png" style="width: 875px; height: 324px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.rank_genes_groups_matrixplot(adata_vis, groupby=&#39;domain&#39;, groups=[&#39;2&#39;, &#39;5&#39;], values_to_plot=&#39;logfoldchanges&#39;, cmap=&#39;RdBu_r&#39;, vmin=-2, vmax=2)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: Groups are not reordered because the `groupby` categories and the `var_group_labels` are different.
categories: 1, 2, 3, etc.
var_group_labels: 2, 5
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Human_Lymph_Node_47_1.png" class="no-scaled-link" src="../_images/notebooks_Human_Lymph_Node_47_1.png" style="width: 645px; height: 304px;" />
</div>
</div>
</section>
<section id="Summarize-the-average-proportion-of-selected-cell-types-in-each-domain">
<h3>Summarize the average proportion of selected cell-types in each domain<a class="headerlink" href="#Summarize-the-average-proportion-of-selected-cell-types-in-each-domain" title="Link to this heading">#</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>gc = [&#39;B_Cycling&#39;, &#39;B_naive&#39;, &#39;B_GC_DZ&#39;, &#39;B_GC_prePB&#39;, &#39;B_preGC&#39;, &#39;T_CD4+_TfH_GC&#39;]
proc.summarize_domain(gc, adata=st_adata)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_Human_Lymph_Node_49_0.png" class="no-scaled-link" src="../_images/notebooks_Human_Lymph_Node_49_0.png" style="width: 1175px; height: 361px;" />
</div>
</div>
</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="MERFISH.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Reveal multiple level of biological heterogeneities in Mouse Hypothalamus by MERFISH-seq</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="DLFPC.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Spatial domain identification on 10x Visium DLFPC data</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2024, Lounan Li
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Performing integrative analysis on scRNA-seq data and 10x Visium data of Human Lymph Node with STEP</a><ul>
<li><a class="reference internal" href="#Loading-packages">Loading packages</a></li>
<li><a class="reference internal" href="#Load-data">Load data</a></li>
<li><a class="reference internal" href="#Initialize-the-STEP-instance">Initialize the STEP instance</a></li>
<li><a class="reference internal" href="#Co-embedding-two-modalities">Co-embedding two modalities</a><ul>
<li><a class="reference internal" href="#Get-and-check-the-co-embedding-results">Get and check the co-embedding results</a></li>
</ul>
</li>
<li><a class="reference internal" href="#Zonation-after-integration">Zonation after integration</a></li>
<li><a class="reference internal" href="#Cell-type-deconvolution-with-obtained-embedding">Cell-type deconvolution with obtained embedding</a><ul>
<li><a class="reference internal" href="#Visualize-the-inferred-cell-type-composition">Visualize the inferred cell-type composition</a></li>
<li><a class="reference internal" href="#Perform-spatially-DE-genes-anlaysis">Perform spatially DE genes anlaysis</a></li>
<li><a class="reference internal" href="#Summarize-the-average-proportion-of-selected-cell-types-in-each-domain">Summarize the average proportion of selected cell-types in each domain</a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="../_static/documentation_options.js?v=5929fcd5"></script>
    <script src="../_static/doctools.js?v=888ff710"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/scripts/furo.js?v=32e29ea5"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=0b406e39"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    </body>
</html>