<!doctype html>
<html class="no-js" lang="en" data-content_root="../">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="../genindex.html" /><link rel="search" title="Search" href="../search.html" /><link rel="next" title="Integrate Human Immune Cell datasets with STEP and generate batch-corrected embeddings and gene expressions." href="scRNA-seq.html" /><link rel="prev" title="Performing integrative analysis on scRNA-seq data and 10x Visium data of Human Lymph Node with STEP" href="Human_Lymph_Node.html" />

    <!-- Generated with Sphinx 7.2.6 and Furo 2024.01.29 -->
        <title>Reveal multiple level of biological heterogeneities in Mouse Hypothalamus by MERFISH-seq - STEP documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo.css?v=135e06be" />
    <link rel="stylesheet" type="text/css" href="../_static/graphviz.css?v=eafc0fe6" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/nbsphinx-code-cells.css" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/furo-extensions.css?v=36a5483c" />
    <link rel="stylesheet" type="text/css" href="../_static/css/custom.css?v=f8d6e08a" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../index.html"><div class="brand">STEP  documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../index.html">
  
  
  <span class="sidebar-brand-text">STEP  documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../autoapi/index.html">Installation</a></li>
<li class="toctree-l1 current has-children"><a class="reference internal" href="../autoapi/index.html#tutorials">Tutorials</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Tutorials</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="DLFPC.html">Spatial domain identification on 10x Visium DLFPC data</a></li>
<li class="toctree-l2"><a class="reference internal" href="Human_Lymph_Node.html">Performing integrative analysis on scRNA-seq data and 10x Visium data of Human Lymph Node with STEP</a></li>
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">Reveal multiple level of biological heterogeneities in Mouse Hypothalamus by MERFISH-seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="scRNA-seq.html">Integrate Human Immune Cell datasets with STEP and generate batch-corrected embeddings and gene expressions.</a></li>
<li class="toctree-l2"><a class="reference internal" href="human_colorectal_cancer.html">Visium HD Human Colorectal Cancer (16 um) cell type clustering &amp; spatial domain identification</a></li>
<li class="toctree-l2"><a class="reference internal" href="mouse_small_intestine.html">Visium HD Mouse Small Intestine (8 um and 16 um) cell type clustering &amp; spatial domain identification</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../autoapi/index.html#api-reference">API Reference</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of API Reference</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../autoapi/step/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" role="switch" type="checkbox"/><label for="toctree-checkbox-3"><div class="visually-hidden">Toggle navigation of step</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3 has-children"><a class="reference internal" href="../autoapi/step/functionality/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.functionality</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" role="switch" type="checkbox"/><label for="toctree-checkbox-4"><div class="visually-hidden">Toggle navigation of step.functionality</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l4 has-children"><a class="reference internal" href="../autoapi/step/functionality/comps/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.functionality.comps</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" role="switch" type="checkbox"/><label for="toctree-checkbox-5"><div class="visually-hidden">Toggle navigation of step.functionality.comps</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l5"><a class="reference internal" href="../autoapi/step/functionality/comps/model_ops/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.functionality.comps.model_ops</span></code></a></li>
<li class="toctree-l5"><a class="reference internal" href="../autoapi/step/functionality/comps/trainer/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.functionality.comps.trainer</span></code></a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/functionality/base/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.functionality.base</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/functionality/cross_funcmodel/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.functionality.cross_funcmodel</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/functionality/sc_funcmodel/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.functionality.sc_funcmodel</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/functionality/st_funcmodel/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.functionality.st_funcmodel</span></code></a></li>
</ul>
</li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../autoapi/step/manager/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.manager</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" role="switch" type="checkbox"/><label for="toctree-checkbox-6"><div class="visually-hidden">Toggle navigation of step.manager</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/manager/save/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.manager.save</span></code></a></li>
</ul>
</li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../autoapi/step/models/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.models</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" role="switch" type="checkbox"/><label for="toctree-checkbox-7"><div class="visually-hidden">Toggle navigation of step.models</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/models/clust/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.models.clust</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/models/deconv/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.models.deconv</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/models/distributions/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.models.distributions</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/models/extension/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.models.extension</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/models/geneformer/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.models.geneformer</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/models/knn_map/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.models.knn_map</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/models/nnls/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.models.nnls</span></code></a></li>
</ul>
</li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../autoapi/step/modules/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.modules</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" role="switch" type="checkbox"/><label for="toctree-checkbox-8"><div class="visually-hidden">Toggle navigation of step.modules</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/modules/decoder/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.modules.decoder</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/modules/gnn/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.modules.gnn</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/modules/transformer/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.modules.transformer</span></code></a></li>
</ul>
</li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../autoapi/step/utils/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.utils</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" role="switch" type="checkbox"/><label for="toctree-checkbox-9"><div class="visually-hidden">Toggle navigation of step.utils</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l4 has-children"><a class="reference internal" href="../autoapi/step/utils/synthetics/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.utils.synthetics</span></code></a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" role="switch" type="checkbox"/><label for="toctree-checkbox-10"><div class="visually-hidden">Toggle navigation of step.utils.synthetics</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l5"><a class="reference internal" href="../autoapi/step/utils/synthetics/pseudo_st/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.utils.synthetics.pseudo_st</span></code></a></li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/utils/dataset/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.utils.dataset</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/utils/gbolt/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.utils.gbolt</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/utils/metrics/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.utils.metrics</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/utils/misc/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.utils.misc</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="../autoapi/step/utils/plotting/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.utils.plotting</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../autoapi/step/integration/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.integration</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../autoapi/step/scmodel/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.scmodel</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="../autoapi/step/stmodel/index.html"><code class="xref py py-mod docutils literal notranslate"><span class="pre">step.stmodel</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="edit-this-page">
  <a class="muted-link" href="https://github.com/SGGb0nd/step/edit/main/docs/notebooks/MERFISH.ipynb" title="Edit this page">
    <svg aria-hidden="true" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentColor" fill="none" stroke-linecap="round" stroke-linejoin="round">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <path d="M4 20h4l10.5 -10.5a1.5 1.5 0 0 0 -4 -4l-10.5 10.5v4" />
      <line x1="13.5" y1="6.5" x2="17.5" y2="10.5" />
    </svg>
    <span class="visually-hidden">Edit this page</span>
  </a>
</div><div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <style>
    .nbinput .prompt,
    .nboutput .prompt {
        display: none;
    }
</style><section id="Reveal-multiple-level-of-biological-heterogeneities-in-Mouse-Hypothalamus-by-MERFISH-seq">
<h1>Reveal multiple level of biological heterogeneities in Mouse Hypothalamus by MERFISH-seq<a class="headerlink" href="#Reveal-multiple-level-of-biological-heterogeneities-in-Mouse-Hypothalamus-by-MERFISH-seq" title="Link to this heading">#</a></h1>
<p>In this tutorial, we will showcae the STEP for revealing multiple levels of biological heterogeneities in the Mouse Hypothalamus by MERFISH-seq. STEP has been designed to shed light on different aspects of cellular complexity: 1. <strong>Cell State Level Heterogeneity</strong>: STEP can capture unique cellular states based on genetic expressions in the mouse hypothalamus. 2. <strong>Spatial Domain Level Heterogeneity</strong>: STEP facilitates understanding of spatial patterns and relationships among cells by smoothing
the cell state embeddings. 3. <strong>End-to-End (e2e) Mode for Direct Spatial Domain Identification</strong>: A streamlined process to directly identify and analyze spatial heterogeneity.</p>
<p>The aim of this tutorial is to guide users through the process of utilizing and understanding the capabilities of STEP, enabling users to dive deep into the biological intricacies of the mouse hypothalamus.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import os
import anndata as ad
import scanpy as sc
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd

from step import scModel, stModel
sc.settings.figdir = &#39;./results/MERFISH&#39;
sc.set_figure_params(dpi_save=300, figsize=(6, 4.5))

st_file = &#39;./data/merfish_animal1.h5ad&#39;
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/projects/82505004-e7a0-445f-ab3c-80d03c91438f/.cache/pypoetry/virtualenvs/step-Ajq_Bw_i-py3.10/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
</pre></div></div>
</div>
<section id="Cell-state-level-heterogeneity">
<h2>Cell state level heterogeneity<a class="headerlink" href="#Cell-state-level-heterogeneity" title="Link to this heading">#</a></h2>
<p>For the identification of cell state level hetorogeneity single-cell resolution data, we use <code class="docutils literal notranslate"><span class="pre">step.scModel</span></code> which will focus on the modeling of transcipts and elimination of the potential batch-effects (not presented in this data), ignoring the spatial context even though the data is spatially resovled.</p>
<section id="Setup-STEP-object">
<h3>Setup STEP object<a class="headerlink" href="#Setup-STEP-object" title="Link to this heading">#</a></h3>
<p>Due to the limitation of the MERFISH technique, there’re only 161 genes across sections. Subsequently, we will use all these genes by setting the paramter <code class="docutils literal notranslate"><span class="pre">n_top_genes=None</span></code> and set the parameter <code class="docutils literal notranslate"><span class="pre">n_modules=20</span></code> to model the fewer gene modules.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>stepc = scModel(
    file_path=st_file,
    n_top_genes=None,
    layer_key=None,
    batch_key=&#39;Bregma&#39;,
    class_key=&#39;Cell_class&#39;,
    module_dim=30,
    hidden_dim=64,
    n_modules=20,
    decoder_type=&#39;zinb&#39;,
    n_dec_hid_layers=1,
)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Using all genes
Using given geneset
Adding count data to layer &#39;counts&#39;
================Dataset Info================
Batch key: Bregma
Class key: Cell_class
Number of Batches: 5
Number of Classes: 15
Gene Expr: (28316, 161)
Batch Label: (28316,)
============================================
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>stepc.adata.obs.head()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Animal_ID</th>
      <th>Animal_sex</th>
      <th>Behavior</th>
      <th>Bregma</th>
      <th>Centroid_X</th>
      <th>Centroid_Y</th>
      <th>Cell_class</th>
      <th>Neuron_cluster_ID</th>
      <th>in_tissue</th>
      <th>n_genes</th>
    </tr>
    <tr>
      <th>Cell_ID</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>6c59b1ec-95df-4f25-882a-ac94d99b87d8</th>
      <td>1</td>
      <td>Female</td>
      <td>Naive</td>
      <td>-0.04</td>
      <td>-3033.401378</td>
      <td>2825.155877</td>
      <td>Endothelial 3</td>
      <td>NaN</td>
      <td>True</td>
      <td>29</td>
    </tr>
    <tr>
      <th>aa7ce410-c485-42c9-9c29-bc1d90a46e00</th>
      <td>1</td>
      <td>Female</td>
      <td>Naive</td>
      <td>-0.04</td>
      <td>-3027.328655</td>
      <td>2955.936449</td>
      <td>Inhibitory</td>
      <td>I-5</td>
      <td>True</td>
      <td>61</td>
    </tr>
    <tr>
      <th>8426329e-077b-4385-bd94-c9fdaeb89bd1</th>
      <td>1</td>
      <td>Female</td>
      <td>Naive</td>
      <td>-0.04</td>
      <td>-3020.512953</td>
      <td>2961.696710</td>
      <td>Inhibitory</td>
      <td>I-5</td>
      <td>True</td>
      <td>54</td>
    </tr>
    <tr>
      <th>a7e4e0de-8baa-4136-8cac-fa9bc759e7e8</th>
      <td>1</td>
      <td>Female</td>
      <td>Naive</td>
      <td>-0.04</td>
      <td>-3017.397490</td>
      <td>2995.915801</td>
      <td>Inhibitory</td>
      <td>I-5</td>
      <td>True</td>
      <td>86</td>
    </tr>
    <tr>
      <th>c55468d5-1b2c-4477-b4ad-3653f5bbda35</th>
      <td>1</td>
      <td>Female</td>
      <td>Naive</td>
      <td>-0.04</td>
      <td>-3005.478613</td>
      <td>2877.528929</td>
      <td>Endothelial 1</td>
      <td>NaN</td>
      <td>True</td>
      <td>40</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>stepc.run(epochs=400,
          batch_size=2048,
          split_rate=0.2,
          tune_epochs=100,
          beta1=0.1,
          beta2=0.01,
         )
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
================Dataset Info================
Batch key: Bregma
Class key: Cell_class
Number of Batches: 5
Number of Classes: 15
Gene Expr: (28316, 161)
Batch Label: (28316,)
============================================
Performing category random split
Training size for -0.24: 4434
Training size for -0.19: 4641
Training size for -0.14: 4740
Training size for -0.09: 4445
Training size for -0.04: 4390
train size: 22650
valid size: 5666
Current Mode: multi_batches: [&#39;gene_expr&#39;, &#39;batch_label&#39;]
Current Mode: multi_batches: [&#39;gene_expr&#39;, &#39;batch_label&#39;]
100%|██████████| 400/400 [05:15&lt;00:00,  1.27epoch/s, kl_loss=2.391, recon_loss=160.094, val_kl_loss=2.473/2.288, val_recon_loss=154.405/153.951]
Current Mode: multi_batches: [&#39;gene_expr&#39;, &#39;batch_label&#39;]
Current Mode: multi_batches: [&#39;gene_expr&#39;, &#39;batch_label&#39;]
Current Mode: multi_batches_with_ct: [&#39;gene_expr&#39;, &#39;class_label&#39;, &#39;batch_label&#39;]
================Dataset Info================
Batch key: Bregma
Class key: Cell_class
Number of Batches: 5
Number of Classes: 15
Gene Expr: (28316, 161)
Class Label: (28316,)
Batch Label: (28316,)
============================================
Performing category random split
Training size for Astrocyte: 2724
Training size for Endothelial 1: 1184
Training size for Endothelial 2: 115
Training size for Endothelial 3: 439
Training size for Ependymal: 1029
Training size for Excitatory: 4629
Training size for Inhibitory: 9895
Training size for Microglia: 475
Training size for OD Immature 1: 809
Training size for OD Immature 2: 25
Training size for OD Mature 1: 190
Training size for OD Mature 2: 922
Training size for OD Mature 3: 11
Training size for OD Mature 4: 37
Training size for Pericytes: 163
train size: 22647
valid size: 5669
Current Mode: multi_batches_with_ct: [&#39;gene_expr&#39;, &#39;class_label&#39;, &#39;batch_label&#39;]
Current Mode: multi_batches_with_ct: [&#39;gene_expr&#39;, &#39;class_label&#39;, &#39;batch_label&#39;]
 32%|███▏      | 32/100 [00:31&lt;01:04,  1.05epoch/s, cl_loss=0.036, kl_loss=0.009, val_cl_loss=0.317/0.233, val_kl_loss=0.010/0.010]Early Stopping triggered
 32%|███▏      | 32/100 [00:31&lt;01:06,  1.03epoch/s, cl_loss=0.036, kl_loss=0.009, val_cl_loss=0.317/0.233, val_kl_loss=0.010/0.010]
Current Mode: multi_batches: [&#39;gene_expr&#39;, &#39;batch_label&#39;]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
EarlyStopping counter: 20 out of 20
EarlyStopping counter: 2 out of 1
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Current Mode: multi_batches_with_ct: [&#39;gene_expr&#39;, &#39;class_label&#39;, &#39;batch_label&#39;]
Current Mode: multi_batches: [&#39;gene_expr&#39;, &#39;batch_label&#39;]
Current Mode: multi_batches: [&#39;gene_expr&#39;, &#39;batch_label&#39;]
Current Mode: multi_batches: [&#39;gene_expr&#39;, &#39;batch_label&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = stepc.adata

sc.pp.neighbors(adata, use_rep=&#39;X_rep&#39;)
sc.tl.umap(adata)
sc.pl.umap(adata, color=[&#39;Bregma&#39;, &#39;Cell_class&#39;])
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_MERFISH_10_0.png" class="no-scaled-link" src="../_images/notebooks_MERFISH_10_0.png" style="width: 1149px; height: 330px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pp.neighbors(adata, use_rep=&#39;X_anchord&#39;)
sc.tl.umap(adata)
sc.pl.umap(adata, color=[&#39;Bregma&#39;, &#39;Cell_class&#39;])
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_MERFISH_11_0.png" class="no-scaled-link" src="../_images/notebooks_MERFISH_11_0.png" style="width: 1149px; height: 330px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from scib_metrics.benchmark import Benchmarker, BioConservation

biocons = BioConservation(nmi_ari_cluster_labels_leiden=True, nmi_ari_cluster_labels_kmeans=False)
# batcons = BatchCorrection(pcr_comparison=False)
bm = Benchmarker(
    adata,
    batch_key=&quot;Bregma&quot;,
    label_key=&quot;Cell_class&quot;,
    embedding_obsm_keys=[&quot;X_rep&quot;, &quot;X_anchord&quot;, &quot;X_pca_corrected&quot;, &quot;X_pca&quot;],
    bio_conservation_metrics=biocons,
    # batch_correction_metrics=batcons,
    n_jobs=-1,
)
bm.benchmark()
bm.plot_results_table(min_max_scale=False)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Computing neighbors: 100%|██████████| 2/2 [00:09&lt;00:00,  4.70s/it]
Embeddings:   0%|<span class="ansi-green-fg">          </span>| 0/2 [00:00&lt;?, ?it/s]
Metrics:   0%|<span class="ansi-blue-fg">          </span>| 0/10 [00:00&lt;?, ?it/s]
Metrics:   0%|<span class="ansi-blue-fg">          </span>| 0/10 [00:00&lt;?, ?it/s, Bio conservation: isolated_labels]WARNING:jax._src.xla_bridge:An NVIDIA GPU may be present on this machine, but a CUDA-enabled jaxlib is not installed. Falling back to cpu.

Metrics:  10%|<span class="ansi-blue-fg">█         </span>| 1/10 [00:02&lt;00:20,  2.30s/it, Bio conservation: isolated_labels]
Metrics:  10%|<span class="ansi-blue-fg">█         </span>| 1/10 [00:02&lt;00:20,  2.30s/it, Bio conservation: nmi_ari_cluster_labels_leiden]
Metrics:  20%|<span class="ansi-blue-fg">██        </span>| 2/10 [00:10&lt;00:46,  5.78s/it, Bio conservation: nmi_ari_cluster_labels_leiden]
Metrics:  20%|<span class="ansi-blue-fg">██        </span>| 2/10 [00:10&lt;00:46,  5.78s/it, Bio conservation: silhouette_label]
Metrics:  30%|<span class="ansi-blue-fg">███       </span>| 3/10 [00:12&lt;00:27,  3.94s/it, Bio conservation: silhouette_label]
Metrics:  30%|<span class="ansi-blue-fg">███       </span>| 3/10 [00:12&lt;00:27,  3.94s/it, Bio conservation: clisi_knn]
Metrics:  40%|<span class="ansi-blue-fg">████      </span>| 4/10 [00:13&lt;00:16,  2.69s/it, Bio conservation: clisi_knn]
Metrics:  40%|<span class="ansi-blue-fg">████      </span>| 4/10 [00:13&lt;00:16,  2.69s/it, Batch correction: silhouette_batch]
Metrics:  50%|<span class="ansi-blue-fg">█████     </span>| 5/10 [00:19&lt;00:19,  3.93s/it, Batch correction: silhouette_batch]
Metrics:  50%|<span class="ansi-blue-fg">█████     </span>| 5/10 [00:19&lt;00:19,  3.93s/it, Batch correction: ilisi_knn]
Metrics:  60%|<span class="ansi-blue-fg">██████    </span>| 6/10 [00:19&lt;00:10,  2.67s/it, Batch correction: ilisi_knn]
Metrics:  60%|<span class="ansi-blue-fg">██████    </span>| 6/10 [00:19&lt;00:10,  2.67s/it, Batch correction: kbet_per_label]
Metrics:  70%|<span class="ansi-blue-fg">███████   </span>| 7/10 [00:35&lt;00:21,  7.01s/it, Batch correction: kbet_per_label]
Metrics:  70%|<span class="ansi-blue-fg">███████   </span>| 7/10 [00:35&lt;00:21,  7.01s/it, Batch correction: graph_connectivity]
Metrics:  80%|<span class="ansi-blue-fg">████████  </span>| 8/10 [00:35&lt;00:14,  7.01s/it, Batch correction: pcr_comparison]
Embeddings:  50%|<span class="ansi-green-fg">█████     </span>| 1/2 [00:36&lt;00:36, 36.03s/it]atch correction: pcr_comparison]
Metrics:   0%|<span class="ansi-blue-fg">          </span>| 0/10 [00:00&lt;?, ?it/s]

Metrics:   0%|<span class="ansi-blue-fg">          </span>| 0/10 [00:00&lt;?, ?it/s, Bio conservation: isolated_labels]
Metrics:  10%|<span class="ansi-blue-fg">█         </span>| 1/10 [00:01&lt;00:16,  1.83s/it, Bio conservation: isolated_labels]
Metrics:  10%|<span class="ansi-blue-fg">█         </span>| 1/10 [00:01&lt;00:16,  1.83s/it, Bio conservation: nmi_ari_cluster_labels_leiden]
Metrics:  20%|<span class="ansi-blue-fg">██        </span>| 2/10 [00:11&lt;00:50,  6.26s/it, Bio conservation: nmi_ari_cluster_labels_leiden]
Metrics:  20%|<span class="ansi-blue-fg">██        </span>| 2/10 [00:11&lt;00:50,  6.26s/it, Bio conservation: silhouette_label]
Metrics:  30%|<span class="ansi-blue-fg">███       </span>| 3/10 [00:12&lt;00:29,  4.17s/it, Bio conservation: silhouette_label]
Metrics:  30%|<span class="ansi-blue-fg">███       </span>| 3/10 [00:12&lt;00:29,  4.17s/it, Bio conservation: clisi_knn]
Metrics:  40%|<span class="ansi-blue-fg">████      </span>| 4/10 [00:13&lt;00:15,  2.60s/it, Bio conservation: clisi_knn]
Metrics:  40%|<span class="ansi-blue-fg">████      </span>| 4/10 [00:13&lt;00:15,  2.60s/it, Batch correction: silhouette_batch]
Metrics:  50%|<span class="ansi-blue-fg">█████     </span>| 5/10 [00:13&lt;00:09,  1.83s/it, Batch correction: silhouette_batch]
Metrics:  50%|<span class="ansi-blue-fg">█████     </span>| 5/10 [00:13&lt;00:09,  1.83s/it, Batch correction: ilisi_knn]
Metrics:  60%|<span class="ansi-blue-fg">██████    </span>| 6/10 [00:13&lt;00:05,  1.27s/it, Batch correction: ilisi_knn]
Metrics:  60%|<span class="ansi-blue-fg">██████    </span>| 6/10 [00:13&lt;00:05,  1.27s/it, Batch correction: kbet_per_label]
Metrics:  70%|<span class="ansi-blue-fg">███████   </span>| 7/10 [00:30&lt;00:18,  6.29s/it, Batch correction: kbet_per_label]
Metrics:  70%|<span class="ansi-blue-fg">███████   </span>| 7/10 [00:30&lt;00:18,  6.29s/it, Batch correction: graph_connectivity]
Embeddings: 100%|<span class="ansi-green-fg">██████████</span>| 2/2 [01:06&lt;00:00, 33.24s/it]atch correction: pcr_comparison]


</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_MERFISH_12_1.png" class="no-scaled-link" src="../_images/notebooks_MERFISH_12_1.png" style="width: 1030px; height: 234px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;plottable.table.Table at 0x7fdb6c7edd20&gt;
</pre></div></div>
</div>
</section>
</section>
<section id="Spatial-domain-level-heterogeneity-by-spatially-smoothing-the-embedding-in-cell-state-level">
<h2>Spatial domain level heterogeneity by spatially smoothing the embedding in cell state level<a class="headerlink" href="#Spatial-domain-level-heterogeneity-by-spatially-smoothing-the-embedding-in-cell-state-level" title="Link to this heading">#</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>model_config = dict(
    n_glayers=4,
    max_neighbors=20,
    variational=True,
)

run_config = dict(
    epochs=400,
    split_rate=0.2,
    n_iterations=800,
    batch_size=1024,
    graph_batch_size=2,
    tune_lr=1e-4,
    beta=1e-4,
    e2e=False,
)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>stepc = stModel(
    file_path=st_file,
    n_top_genes=None,
    batch_key=&#39;Bregma&#39;,
    filtered=True,
    layer_key=None,
    log_transformed=True,
    coord_keys=[&#39;Centroid_X&#39;, &#39;Centroid_Y&#39;],
    module_dim=30,
    hidden_dim=64,
    n_modules=20,
    decoder_type=&#39;zinb&#39;,
    edge_clip=None,
    dispersion=&#39;gene&#39;,
    use_earlystop=True,
    **model_config,
)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Using all genes
Using given geneset
Adding count data to layer &#39;counts&#39;
Dataset Done
================Dataset Info================
Batch key: Bregma
Class key: None
Number of Batches: 5
Number of Classes: None
Gene Expr: (28316, 161)
Batch Label: (28316,)
============================================
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>stepc.run(**run_config)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Training with 2 stages pattern: 1/2
Performing global random split
100%|██████████| 400/400 [03:48&lt;00:00,  1.75epoch/s, kl_loss=0.880, recon_loss=146.231, val_kl_loss=0.884, val_recon_loss=147.476]
Training graph with multiple batches
Training with 2 stages pattern: 2/2
100%|██████████| 800/800 [07:45&lt;00:00,  1.72epoch/s, kl_loss=0.012, recon_loss=212.722]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = stepc.adata

sc.pp.neighbors(adata, use_rep=&#39;X_unsmoothed&#39;)
sc.tl.umap(adata)
sc.pl.umap(adata, color=[&#39;Bregma&#39;, &#39;Cell_class&#39;])
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_MERFISH_17_0.png" class="no-scaled-link" src="../_images/notebooks_MERFISH_17_0.png" style="width: 1149px; height: 330px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>stepc.cluster(adata=adata, use_rep=&#39;X_smoothed&#39;, n_clusters=8)
sc.pp.neighbors(adata, use_rep=&#39;X_smoothed&#39;)
sc.tl.umap(adata)
sc.pl.umap(adata, color=[&#39;Bregma&#39;, &#39;Cell_class&#39;, &#39;domain&#39;])
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_MERFISH_18_0.png" class="no-scaled-link" src="../_images/notebooks_MERFISH_18_0.png" style="width: 1353px; height: 330px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>stepc.spatial_plot(color=&#39;domain&#39;, with_images=False, size=40)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Title not provided, or length does not match the number of batches. Using feature names as title.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_MERFISH_19_1.png" class="no-scaled-link" src="../_images/notebooks_MERFISH_19_1.png" style="width: 450px; height: 326px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_MERFISH_19_2.png" class="no-scaled-link" src="../_images/notebooks_MERFISH_19_2.png" style="width: 450px; height: 326px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_MERFISH_19_3.png" class="no-scaled-link" src="../_images/notebooks_MERFISH_19_3.png" style="width: 450px; height: 326px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_MERFISH_19_4.png" class="no-scaled-link" src="../_images/notebooks_MERFISH_19_4.png" style="width: 450px; height: 326px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_MERFISH_19_5.png" class="no-scaled-link" src="../_images/notebooks_MERFISH_19_5.png" style="width: 450px; height: 326px;" />
</div>
</div>
</section>
<section id="E2E-mode-for-direct-spatial-domain-identification">
<h2>E2E mode for direct spatial domain identification<a class="headerlink" href="#E2E-mode-for-direct-spatial-domain-identification" title="Link to this heading">#</a></h2>
<section id="Model-config-&amp;-running-config">
<h3>Model config &amp; running config<a class="headerlink" href="#Model-config-&-running-config" title="Link to this heading">#</a></h3>
<p>First, set <code class="docutils literal notranslate"><span class="pre">e2e=True</span></code> to enable the e2e mode. In e2e mode, these parameters used in the non e2e mode will be ignored: - <code class="docutils literal notranslate"><span class="pre">epochs</span></code> - <code class="docutils literal notranslate"><span class="pre">batch_size</span></code> - <code class="docutils literal notranslate"><span class="pre">split_rate</span></code></p>
<p>Instead, a graph sampling strategy will be used to train the model.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>model_config = dict(
    n_glayers=4,
    max_neighbors=30,
    variational=True,
)

run_config = dict(
    n_iterations=2000,
    graph_batch_size=1,
    beta=1e-2,
    e2e=True,
    contrast=True,
)
</pre></div>
</div>
</div>
</section>
<section id="Load-data-and-add-annotations-of-zones">
<h3>Load data and add annotations of zones<a class="headerlink" href="#Load-data-and-add-annotations-of-zones" title="Link to this heading">#</a></h3>
<div class="admonition note">
<p><strong>Note</strong>: The annotations of zones is obtained from the <a class="reference external" href="https://zhengli09.github.io/BASS-Analysis/MERFISH.html">tutorial of BASS</a></p>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = sc.read_h5ad(st_file)

for batch in os.listdir(&#39;./data/merfish_anno/&#39;):
    anno = pd.read_csv(f&#39;./data/merfish_anno/{batch}&#39;, index_col=0)
    batch_id = batch[:-4][6:]
    adata.obs.loc[adata.obs[&#39;Bregma&#39;] == float(batch_id), &#39;annotation&#39;] = anno[&#39;x&#39;].values
</pre></div>
</div>
</div>
</section>
<section id="Set-up-STEP-object">
<h3>Set up STEP object<a class="headerlink" href="#Set-up-STEP-object" title="Link to this heading">#</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>stepc = stModel(
    adata=adata,
    n_top_genes=None,
    geneset_to_use=adata.var_names,
    batch_key=&#39;Bregma&#39;,
    filtered=True,
    layer_key=None,
    log_transformed=True,
    coord_keys=[&#39;Centroid_X&#39;, &#39;Centroid_Y&#39;],
    module_dim=30,
    hidden_dim=64,
    n_modules=20,
    decoder_type=&#39;zinb&#39;,
    edge_clip=None,
    **model_config,
)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Adding count data to layer &#39;counts&#39;
Dataset Done
Modeling with 1 components
{&#39;variational&#39;: True, &#39;input_dim&#39;: 161, &#39;hidden_dim&#39;: 64, &#39;module_dim&#39;: 30, &#39;decoder_input_dim&#39;: None, &#39;n_modules&#39;: 20, &#39;edge_clip&#39;: None, &#39;decoder_type&#39;: &#39;zinb&#39;, &#39;n_glayers&#39;: 4}
</pre></div></div>
</div>
</section>
<section id="Generate-spatially-smoothed-embeddings-and-cluster-them-to-obtain-spatial-domains">
<h3>Generate spatially smoothed embeddings and cluster them to obtain spatial domains<a class="headerlink" href="#Generate-spatially-smoothed-embeddings-and-cluster-them-to-obtain-spatial-domains" title="Link to this heading">#</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>stepc.run(**run_config)
adata = stepc.adata
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Training with e2e pattern
Training graph with multiple batches
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 2000/2000 [04:22&lt;00:00,  7.60step/s, recon_loss=370.799, kl_loss=0.428, contrast_loss=0.011, val_recon_loss=-, val_kl_loss=-, val_contrast_loss=-]
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>stepc.cluster(adata, use_rep=&#39;X_smoothed&#39;, n_clusters=8,)
</pre></div>
</div>
</div>
</section>
<section id="Visualization">
<h3>Visualization<a class="headerlink" href="#Visualization" title="Link to this heading">#</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>for batch in adata.obs[&#39;Bregma&#39;].unique():
    _adata = adata[adata.obs[&#39;Bregma&#39;] == batch]
    plt.title(batch)
    plt.scatter(x=_adata.obs[&#39;Centroid_X&#39;],
            y=_adata.obs[&#39;Centroid_Y&#39;],
            c=_adata.obs[&#39;domain&#39;].values.astype(int),
            s=5, cmap=&#39;tab20&#39;)
    plt.show()
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_MERFISH_33_0.png" class="no-scaled-link" src="../_images/notebooks_MERFISH_33_0.png" style="width: 445px; height: 331px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_MERFISH_33_1.png" class="no-scaled-link" src="../_images/notebooks_MERFISH_33_1.png" style="width: 437px; height: 331px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_MERFISH_33_2.png" class="no-scaled-link" src="../_images/notebooks_MERFISH_33_2.png" style="width: 440px; height: 331px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_MERFISH_33_3.png" class="no-scaled-link" src="../_images/notebooks_MERFISH_33_3.png" style="width: 455px; height: 331px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_MERFISH_33_4.png" class="no-scaled-link" src="../_images/notebooks_MERFISH_33_4.png" style="width: 452px; height: 331px;" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata.obs[&#39;annotation&#39;] = adata.obs[&#39;annotation&#39;].astype(&#39;category&#39;)
</pre></div>
</div>
</div>
</section>
<section id="Metrics-for-evaluation-of-the-obtained-spatial-domains-compared-to-the-annotations">
<h3>Metrics for evaluation of the obtained spatial domains compared to the annotations<a class="headerlink" href="#Metrics-for-evaluation-of-the-obtained-spatial-domains-compared-to-the-annotations" title="Link to this heading">#</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from sklearn.metrics import adjusted_rand_score

for batch in adata.obs[&#39;Bregma&#39;].unique():
    _adata = adata[adata.obs[&#39;Bregma&#39;] == batch]
    ari = adjusted_rand_score(_adata.obs[&#39;domain&#39;].astype(int), _adata.obs[&#39;annotation&#39;].cat.codes)
    print(f&quot;{batch}: {ari}&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
-0.04: 0.5337230222364567
-0.09: 0.6226701393120998
-0.14: 0.49295132575539485
-0.19: 0.6502377274236062
-0.24: 0.6812873877486072
</pre></div></div>
</div>
</section>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="scRNA-seq.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Integrate Human Immune Cell datasets with STEP and generate batch-corrected embeddings and gene expressions.</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="Human_Lymph_Node.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Performing integrative analysis on scRNA-seq data and 10x Visium data of Human Lymph Node with STEP</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2024, Lounan Li
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Reveal multiple level of biological heterogeneities in Mouse Hypothalamus by MERFISH-seq</a><ul>
<li><a class="reference internal" href="#Cell-state-level-heterogeneity">Cell state level heterogeneity</a><ul>
<li><a class="reference internal" href="#Setup-STEP-object">Setup STEP object</a></li>
</ul>
</li>
<li><a class="reference internal" href="#Spatial-domain-level-heterogeneity-by-spatially-smoothing-the-embedding-in-cell-state-level">Spatial domain level heterogeneity by spatially smoothing the embedding in cell state level</a></li>
<li><a class="reference internal" href="#E2E-mode-for-direct-spatial-domain-identification">E2E mode for direct spatial domain identification</a><ul>
<li><a class="reference internal" href="#Model-config-&amp;-running-config">Model config &amp; running config</a></li>
<li><a class="reference internal" href="#Load-data-and-add-annotations-of-zones">Load data and add annotations of zones</a></li>
<li><a class="reference internal" href="#Set-up-STEP-object">Set up STEP object</a></li>
<li><a class="reference internal" href="#Generate-spatially-smoothed-embeddings-and-cluster-them-to-obtain-spatial-domains">Generate spatially smoothed embeddings and cluster them to obtain spatial domains</a></li>
<li><a class="reference internal" href="#Visualization">Visualization</a></li>
<li><a class="reference internal" href="#Metrics-for-evaluation-of-the-obtained-spatial-domains-compared-to-the-annotations">Metrics for evaluation of the obtained spatial domains compared to the annotations</a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="../_static/documentation_options.js?v=5929fcd5"></script>
    <script src="../_static/doctools.js?v=888ff710"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/scripts/furo.js?v=32e29ea5"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=0b406e39"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    </body>
</html>